runtime-modules/content/src/lib.rs

```diff
@@ -2333,6 +2423,101 @@ impl<T: Trait> Module<T> {
         Ok(())
     }
 
+    fn execute_delete_video_mutation(
+        sender: &T::AccountId,
+        channel_id: T::ChannelId,
+        video_id: T::VideoId,
+        video: &Video<T>,
+        assets_to_remove: &BTreeSet<DataObjectId<T>>,
+    ) -> DispatchResult {
+        // bloat bond logic: channel owner is refunded
+        video
+            .video_post_id
+            .as_ref()
+            .map(|video_post_id| {
+                Self::video_deletion_refund_logic(&sender, &video_id, &video_post_id)
+            })
+            .transpose()?;
+
+        if !assets_to_remove.is_empty() {
+            Storage::<T>::delete_data_objects(
+                sender.clone(),
+                Self::bag_id_for_channel(&channel_id),
+                assets_to_remove.clone(),
+            )?;
+        }
+
+        // Remove video
+        VideoById::<T>::remove(video_id);
+
+        // Remove all comments related
+        <VideoPostById<T>>::remove_prefix(video_id);
+
+        // Update corresponding channel
+        // Remove recently deleted video from the channel
+        ChannelById::<T>::mutate(channel_id, |channel| {
+            channel.num_videos = channel.num_videos.saturating_sub(1)
+        });
+
+        Ok(())
+    }
+
+    fn ensure_channel_bag_can_be_dropped(
+        dynamic_bag_id: &storage::DynamicBagId<T>,
+        num_objects_to_delete: u64,
+    ) -> Result<(bool, BTreeSet<DataObjectId<T>>), DispatchError> {
+        let bag_id = storage::BagIdType::from(dynamic_bag_id.clone());
+
+        if let Ok(bag) = T::DataObjectStorage::ensure_bag_exists(&bag_id) {
+            // channel has a dynamic bag associated
+            // ensure that bag size provided is valid
+            ensure!(
+                bag.objects_number == num_objects_to_delete,
+                Error::<T>::InvalidBagSizeSpecified
+            );
+
+            // construct collection of assets to be removed
+            let assets_to_remove = T::DataObjectStorage::get_data_objects_id(&bag_id);
+
+            if !assets_to_remove.is_empty() {
+                Storage::<T>::can_delete_dynamic_bag_with_objects(&dynamic_bag_id)?;
```

Comment:
I factored out this part of the code from `delete_channel`, from what I can tell one reason as to why `can_delete_dynamic_bag_with_objects` is needed here is because it ensures `InsufficientTreasuryBalance` will not be thrown during `delete_dynamic_bag` at https://github.com/Lezek123/substrate-runtime-joystream/blob/content-directory-new-permissions-model/runtime-modules/content/src/lib.rs#L2562 (which happens after `delete_data_objects`, so that would be an unsafe mutation).

Reply:
Replaced by introducing new `delete_dynamic_bag_with_objects` function in https://github.com/Joystream/joystream/pull/3366/commits/4aa4dac265b5e8a12d6e3428a2f16f30977175e3